### DISCUSSION TITLE
DQD, DashboardExporter, and AresIndexer -- has anyone done the entire workflow?

### DISCUSSION THREAD

**Discussion by heschmidt04** [chorus-ai/StandardsModule#86] [Results Sharing Questions] [answered]
I'll put my initial notes below. 
This assumes you've set up RStudio, created a GITHUB_PAT token and put it in your local .Renviron file and installed R packages for these 3 tools. 

Note: These outputs are in first stages as our whole ETL workflow is WIP and our cohort OMOP data is not all created yet. 

**TL;DR Specify the cdmVersion! Defaults to Version 5.3 which has two different names for 2 columns and will fail if your OMOP is Version 5.4** 

**Two questions I have for this entire workflow**
1. Anyone else using Snowflake? 
2. Anyone else have AresIndexer errors? What is correct R code to use? What are the steps to setup and run AresIndexer locally on a Mac laptop? What does success look like for AresIndexer? 

**Summary:** 
- My main challenge is that AresIndexer using the R database connect details package doesn't know how to connect to Snowflake properly. 
- DQD does run and DashboardExporter runs using the db connect details R package. Dashboard Exporter tells me I don't have data for those analysis ids, yet :D so I will have to wait to know what success looks like for that output. 

References:

- [Reference Achilles office hours](https://drive.google.com/drive/folders/1b2EhWZ08OvEjFj5A0uCAbR0SsbQan7e6?usp=drive_link)

- [Reference Ares ETL office hours](https://drive.google.com/drive/folders/19gv_n-WFqoFAM-1MHNpRa5heEUbrjoOg?usp=drive_link)

- [Reference Achilles/DQD submission office hours](https://drive.google.com/drive/folders/1LAmtp7kygnqr1vuGvKS3wmTvbn5gmNZ_?usp=sharing)

### Here are my notes from a Snowflake perspective for Achilles and DashboardExport

- If folks need the DQD R code to run DQD it. 
- This assumes packages and libraries are in place for R
- This connection information is Specific to Snowflake and external browser authentication using SSL and SAML

```
library(DataQualityDashboard)
resultsDatabaseSchema <-"schema" # the fully qualified database schema name of the results schema (that you can write to) 
cdmSourceName <- "Snowflake-MGB-OMOP" # a human readable name for your CDM source
cdmVersion <- "5.4" # the CDM version you are targeting. Currently supports 5.2, 5.3, and 5.4


# determine how many threads (concurrent SQL sessions) to use
numThreads <- 1 # on Redshift, 3 seems to work well

# specify if you want to execute the queries or inspect them
sqlonly <- FALSE # set to TRUE if you just want to get the SQL scripts and not actually run the queries

sqlonlyIncrementalInsert <- FALSE # set to TRUE if you want the generated SQL queries to calculate DQD 
sqlOnlyUnionCount <- 1  # in sqlonlyIncrementalInsert mode, the number of check sals to union in a sin

# write results to tahley
writeToTable <- TRUE # set to FALSE if you want to skip writing to a SQL table in the results schema

# specify the name of the results table (used when writeToTable = TRUE and when sqlOnlyIncrementalInsert = TRUE)
writeTableName <- "dq_dashboard_results"
# write results to a csv file?
writeToCsv <- FALSE # set to FALSE if you want to skip writing to csv file

csvFile <-""		# 

# which DQ check levels to run
checkLevels <- c("TABLE", "FIELD", "CONCEPT")
# which DQ Checks to run? ---
checkNames <- c() # Names can be found in Github repo under inst/csv/OMOP_CDM_v5.3_Check_Descriptions.csv

tablesToExclude <- c("CONCEPT", "VOCABULARY", "CONCEPT_ANCESTOR", "CONCEPT_RELATIONSHIP", "CONCEPT_CLASS", "CONCEPT_SYNONYM", "RELATIONSHIP", "DOMAIN") 

# logging type ---------------------
verboseMode <- TRUE # set to FALSE if you don't want the logs to be printed to the console

# where should the results and logs go? ---------------
outputFolder <- "$HOME_DIR/dqd"
outputFile <- "dqd-results.json"

connectionDetails <- DatabaseConnector::createConnectionDetails(
  dbms = "snowflake",
  user = "redacted",
  password = "password",
  pathToDriver = "/path-to-snowflake-java-jar/java_drivers",
  connectionString = "jdbc:snowflake://snowflake-host.com:443?ROLE=snowflake-user-role-saml&authenticator=externalbrowser&database=omop-database"
)


# Run the checks 
library(DataQualityDashboard::executeDqChecks (connectionDetails - connectionDetails,
                                       cdmDatabaseSchema = cdmDatabaseSchema,
                                       resultsDatabaseSchema = resultsDatabaseSchema,
                                       cdmSourceName = cdmSourceName,
                                       numThreads = numThreads,
                                       sqlonly = sqlOnly,
                                       sqlOnlyUnionCount = sqlOnlyUnionCount,
                                       sqlOnlyIncrementalInsert = sqlOnlyIncrementalInsert,
                                       outputFolder = outputFolder,
                                       verboseMode = verboseMode,
                                       writeToTable = writeToTable,
                                       writeToCsv = writeToCsv,
                                       CsvFile = csvFile,
                                       checkLevels = checkLevels,
                                       tablesToExclude = tablesToExclude,
                                       checkNames = checkNames)

```
#### Running through R - Ran DQD and created the file dqd-results.json in output folder

    Ran Achilles in Snowflake with errors on dropping the temp tables (which should be in memory). Solved with `sqlOnly = True` and `cdmVersion = 5.4`

    Example log error: 
		Done. Achilles results can now be found in schema 
		Done. Achilles results can now be found in schema 
		Dropping scratch Achilles tables from schema #  
		Dropping scratch Achilles tables from schema #
		Connecting using Snowflake driver
		Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser 		windows, or your OS settings. Press CTRL+C to abort and try again...
		Drop Achilles Scratch Table -- ERROR (Error in `.createErrorReport()`:
		! Error executing SQL:
		net.snowflake.client.jdbc.SnowflakeSQLException: Cannot perform DROP. This session does not have a current schema. Call 'USE SCHEMA', or use a qualified name.
		An error report has been created at  $HOME_DIR/errorReportSql.txt
		)

-   In that error scenario above there is an achilles_results_dist, not achilles_results
-   There is a log file for that error scenario which states what was run -- $HOME_DIR/achilles-output/log_achilles.txt
-   The Dashboard export https://github.com/darwin-eu/DashboardExport fails to find the ACHILLES_RESULTS table because there is only achilles_results_dist.   
-   Therefore it doesn't run if missing data or lack of achilles_results table, so no CSV output file in the output folder which is what DashboardExporter is supposed to create one CSV of the achilles tables data. 

Running the SQL file from sqlOnly = True 
Running DQD without cdmVersion set will default to version 5.3 and for a 5.4 OMOP throw 2-3 errors on column name around line 7280 - differences between discharging and discharged and admitting and admitted in column names.
Snowflake requires commenting out the last two drop tables and adding "create or replace table" syntax for the final ACHILLES 2 tables so the data is recreated. 

### Code used 
Main code run after installing both packages and referencing them with Library() call. 

```
Achilles::achilles(
  connectionDetails = connectionDetails, 
  cdmDatabaseSchema = "schema",
  resultsDatabaseSchema = "schema", 
  outputFolder = "/path/to/output/dir/achilles-output",
  sqlOnly = TRUE
  # For running only the minimally required Achilles analyses:
  # analysisIds = DashboardExport::getRequiredAnalysisIds()
)

DashboardExport::dashboardExport(
  connectionDetails = connectionDetails,
  cdmDatabaseSchema = "workspace.schema",
  resultsDatabaseSchema = "workspace.schema",
  outputFolder = "/path/to/output/dir/darwin-dashboard-ouput",
  # databaseId = databaseId, # This won't work for Snowflake as it has role, workspace, schema, etc. 
  smallCellCount = 5
)
```

### DashboardExport

These are the missing ids at the moment from running the 2nd block of code above
```
2024-02-02 10:00:14 Missing Achilles analysis ids in result tables: 101, 102, 103, 105, 108, 110, 111, 113, 117, 400, 401, 403, 405, 420, 700, 701, 703, 705, 720.
2024-02-02 10:00:14 Please rerun Achilles including above analyses.

```
Further data splunking - i.e. run the sql for the temp table in that analysis section - the SQL shows that we don't have data there yet for those result tables which matches with the fact that it is WIP so no 1 lump CSV file of all Achilles results, yet.

**Comment by jshoughtaling** [comment, marked as answer]
@heschmidt04 - thank you for the detailed description of your efforts to run these processes! I have a few comments in response, and perhaps we can discuss in further detail offline.

1. If you run DQD in SQLOnly mode, and set sqlonlyIncrementalInsert = `TRUE`, you can execute the generated SQL queries directly against your Snowflake instance outside of R and populate a `dqdashboard_results` table with all DQD results in that instance. We are taking this approach at Tufts and I added a function to the DQD repo that exports that [database table to a json file](https://github.com/OHDSI/DataQualityDashboard/blob/main/R/writeDBResultsTo.R) for use later on in Ares.
2. Executing the DashboardExport package from Darwin-EU is not necessary for the CHoRUS results delivery process flow. I believe you may be looking for the Achilles function that [exports Achilles results for Ares](https://github.com/OHDSI/Achilles/blob/main/R/exportToAres.R). 
3. In order to produce an AresIndex that can eventually be visualized by the Ares web application, you will need the following:
    - Achilles executed from start to finish with no errors (I see you have an error in dropping scratch tables)
    - DQD executed from start to finish with no errors, producing a json-version of the output
    - Achilles Ares Export executed from start to finish with no errors (this produces per-concept reports that make up the bulk of the index files)
    - Achilles `performTemporalCharacterization` executed to compute additional temporal information per concept where relevant/possible
    - AresIndexer `augmentConceptFiles` to add the temporal information into the per-concept files generated by the Achilles Ares Export
    - **NOTE:** There are other functions in the AresIndexer package (e.g. `buildNetworkIndex`, `buildDataQualityIndex`, and `buildNetworkUnmappedSourceCodeIndex`) that you *DO NOT* need to execute unless you are planning to deploy the Ares web application locally. These construct files that aggregate information across individual source release indexes (i.e. to allow you to compare multiple versions, or multiple OMOP data sets), and we will execute them centrally across all indexes.

**Reply by heschmidt04** [reply]
I'll go back to the exportToAres.R and go from there. And thankyou for the example and context for https://github.com/OHDSI/DataQualityDashboard/blob/main/R/writeDBResultsTo.R#L22 
I'll follow up here so folks understand what I did for workflow. Thanks!

**Reply by jshoughtaling** [reply]
@heschmidt04 - excellent, thank you! What you will need to produce and deliver in the end should look like the following (the `exportToAres` function should produce this directory structure for you):

```
20240113 -> You will ultimately need to zip and upload this release-date directory
├── achilles-performance.csv
├── cdmsource.csv
├── concepts
│   ├── condition_era
│   │   ├── concept_132277.json
│   │   ├── ...
│   │   └── concept_81988.json
│   ├── condition_occurrence
│   │   ├── concept_132277.json
│   │   ├── ...
│   │   └── concept_81931.json
│   ├── device_exposure
│   │   ├── concept_2614781.json
│   │   ├── ...
│   │   └── concept_45971500.json
│   ├── drug_era
│   │   ├── concept_1000560.json
│   │   ├── ...
│   │   └── concept_998415.json
│   ├── drug_exposure
│   │   ├── concept_1000560.json
│   │   ├── ...
│   │   └── concept_998415.json
│   ├── measurement
│   │   ├── concept_1001488.json
│   │   ├── ...
│   │   └── concept_914038.json
│   ├── observation
│   │   ├── concept_1340204.json
│   │   ├── ...
│   │   └── concept_903652.json
│   ├── procedure_occurrence
│   │   ├── concept_1389619.json
│   │   ├── ...
│   │   └── concept_927192.json
│   ├── visit_detail
│   │   ├── concept_38004269.json
│   │   ├── ...
│   │   └── concept_9203.json
│   └── visit_occurrence
│       ├── concept_32036.json
│       ├── ...
│       └── concept_9203.json
├── datadensity-concepts-per-person.csv
├── datadensity-domains-per-person.csv
├── datadensity-records-per-person.csv
├── datadensity-total.csv
├── death.json
├── domain-drug-stratification.csv
├── domain-issues.csv
├── domain-summary-condition_era.csv
├── domain-summary-condition_occurrence.csv
├── domain-summary-device_exposure.csv
├── domain-summary-drug_era.csv
├── domain-summary-drug_exposure.csv
├── domain-summary-measurement.csv
├── domain-summary-observation.csv
├── domain-summary-procedure_occurrence.csv
├── domain-summary-provider.csv
├── domain-summary-visit_detail.csv
├── domain-summary-visit_occurrence.csv
├── domain-visit-stratification.csv
├── dq-result.json
├── dq-result_camel.json [possibly not needed depending on your approach]
├── metadata.csv
├── observationperiod.json
├── person.json
├── quality-completeness.csv
├── records-by-domain.csv
└── temporal-characterization.csv
```

**Comment by del42** [comment]
Do we need to run Achilles first , then DQD? @jshoughtaling 
For context, we added legacy EHR data goes back till 2010, then run the DQD. We surprised to get 97. Seems suspicious. So we are just wondering if there is order of things or did we miss something before DQD, etc.

**Comment by heschmidt04** [comment]
Can you do a code walkthrough of
https://github.com/OHDSI/Achilles/blob/main/R/exportToAres.R and how it
gets run?
It's not clear to me after the 2 R libraries are loaded and the connection
details and variables are set how the rest of the code is instantiated.

Thanks
Heidi

On Thu, Feb 8, 2024 at 1:53 PM Jared Houghtaling ***@***.***>
wrote:

> @heschmidt04 <https://github.com/heschmidt04> - excellent, thank you!
> What you will need to produce and deliver in the end should look like the
> following:
>
> 20240113 -> You will ultimately need to zip and upload this release-date directory
> ├── achilles-performance.csv
> ├── cdmsource.csv
> ├── concepts
> │   ├── condition_era
> │   │   ├── concept_132277.json
> │   │   ├── ...
> │   │   └── concept_81988.json
> │   ├── condition_occurrence
> │   │   ├── concept_132277.json
> │   │   ├── ...
> │   │   └── concept_81931.json
> │   ├── device_exposure
> │   │   ├── concept_2614781.json
> │   │   ├── ...
> │   │   └── concept_45971500.json
> │   ├── drug_era
> │   │   ├── concept_1000560.json
> │   │   ├── ...
> │   │   └── concept_998415.json
> │   ├── drug_exposure
> │   │   ├── concept_1000560.json
> │   │   ├── ...
> │   │   └── concept_998415.json
> │   ├── measurement
> │   │   ├── concept_1001488.json
> │   │   ├── ...
> │   │   └── concept_914038.json
> │   ├── observation
> │   │   ├── concept_1340204.json
> │   │   ├── ...
> │   │   └── concept_903652.json
> │   ├── procedure_occurrence
> │   │   ├── concept_1389619.json
> │   │   ├── ...
> │   │   └── concept_927192.json
> │   ├── visit_detail
> │   │   ├── concept_38004269.json
> │   │   ├── ...
> │   │   └── concept_9203.json
> │   └── visit_occurrence
> │       ├── concept_32036.json
> │       ├── ...
> │       └── concept_9203.json
> ├── datadensity-concepts-per-person.csv
> ├── datadensity-domains-per-person.csv
> ├── datadensity-records-per-person.csv
> ├── datadensity-total.csv
> ├── death.json
> ├── domain-drug-stratification.csv
> ├── domain-issues.csv
> ├── domain-summary-condition_era.csv
> ├── domain-summary-condition_occurrence.csv
> ├── domain-summary-device_exposure.csv
> ├── domain-summary-drug_era.csv
> ├── domain-summary-drug_exposure.csv
> ├── domain-summary-measurement.csv
> ├── domain-summary-observation.csv
> ├── domain-summary-procedure_occurrence.csv
> ├── domain-summary-provider.csv
> ├── domain-summary-visit_detail.csv
> ├── domain-summary-visit_occurrence.csv
> ├── domain-visit-stratification.csv
> ├── dq-result.json
> ├── dq-result_camel.json [possibly not needed depending on your approach]
> ├── metadata.csv
> ├── observationperiod.json
> ├── person.json
> ├── quality-completeness.csv
> ├── records-by-domain.csv
> └── temporal-characterization.csv
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/chorus-ai/StandardsModule/discussions/86#discussioncomment-8411235>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYR4GN7REAF7YGFQHU5INDYSUNKFAVCNFSM6AAAAABCXAXZ2SVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DIMJRGIZTK>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
"Nothing is for itself, but each thing partaking of the other is living its
greatest possibility . . ." -Robert Henri

**Comment by heschmidt04** [comment]
I must be missing something. I ran the code and it appears to give me an error 

- The dq-results.json is in the outputPath directory for reference. 

- The Achilles Results table is in the Snowflake database in the cdmDatabaseSchema

- All Schema variables are the same as cdmDatabaseSchema 

Code run 

> library(DatabaseConnector)
> library(Achilles)
> connectionDetails <- createConnectionDetails(
>   dbms = "snowflake",
>   user = "redacted",
>   password = "redacted",
>   pathToDriver = "/path-to-snowflake-drivers/java_drivers",
>   connectionString = "jdbc:snowflake://snowflakecomputing.com:443?ROLE=ROLE&authenticator=externalbrowser&database=DATABASE"
> )

> cdmDatabaseSchema = "schema-redacted"
> resultsDatabaseSchema = "schema-redacted"
> vocabDatabaseSchema = "schema-redacted"
> nativeSchema <- "schema-redacted"
> outputPath = "/path-to-local/ares-data"
> cdmVersion <- "5.4" 
> cdmSourceName <- "Snowflake-MGB-OMOP"
> Achilles::exportToAres(
   connectionDetails = connectionDetails,
   cdmDatabaseSchema = cdmDatabaseSchema,
   resultsDatabaseSchema = resultsDatabaseSchema,
   vocabDatabaseSchema = vocabDatabaseSchema,
   outputPath = outputPath
 )
Connecting using Snowflake driver
Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...
[1] "processing AO export to /path-to-local/ares-data/MGB_Epic_Clarity/20230831"

Error returned 

> Error in aggregate.data.frame(as.data.frame(x), ...) : 
>   no rows to aggregate`

**Comment by heschmidt04** [comment]
Update on AresIndexer part of the equation.  Thanks Jared! 
Using https://github.com/chorus-ai/chorus-container-apps/blob/main/etl/src/etl/ares.R or another R code was not the issue. 

FYI: Essentially in Snowflake the constraints, fk, pk are not enforced or implemented. 
 
Therefore if there isn't an OMOP observation period table entry for the person id in the OMOP Person table then this Pandas dataframe error will be thrown. 

This github issue for Achilles mentions it, yet doesn't explain the "why" of it. 
https://github.com/OHDSI/Achilles/issues/676

Hope this helps other folks who are developing their ETL in stages and run into this. 
Aside: The errors that appear in the output directory will be under the output directory defined in R code under the CDM_SOURCE_ABBREVIATION/CDM_RELEASE_DATE column values your site put as data into the OMOP CDM_SOURCE table 

Best,
Heidi

**Comment by heschmidt04** [comment]
New issue. Further along ; which is good :D 
I opened up an Ares Exporter issue in Ares Github for my Snowflake numeric type error I got when exporting. I have the results in tables if needed. 

The github issue filed is https://github.com/OHDSI/Ares/issues/366
